File: Project Part 1/TaskB.rb
# Class for book items with necessary attributes and an initializer
class Book 
  attr_accessor :isbn, :author, :title, :publisher

  def initialize(isbn, first, last, title, publisher)
    @isbn = isbn
    @author = first + " " + last
    @title = title
    @publisher = publisher
  end
end

# Function to sort the list of books and write them to the textfile
def sortBooks(bookList, &block)

  # Loop for bubble sort
  loop do
    unsorted = false

    # Iterate through list for each loop
    for i in 0..bookList.size - 2 do

      # Make code block call and do necessary comparisons and swaps
      compare = block.call(bookList[i], bookList[i + 1])
      if compare == false then
        bookList[i], bookList[i + 1] = bookList[i + 1], bookList[i]
        unsorted = true
      end
    end

    # Check if list still needs sorting
    if unsorted == false then
      break
    end
  end

  # Write sorted list back to the textfile
  File.open("textfile.txt", "w") do |file|
    for i in 0..bookList.size - 1 do
      file.write(bookList[i].isbn.to_s + ", " + bookList[i].author + ", " + bookList[i].title + ", " + bookList[i].publisher + "\n")
    end
  end
end

# Declare array for book items and words to read from the textfile
bookList=[]
words=[]

# Open file and read all 10000 books to put into the array
File.open("textfile.txt", "r") do |file|
  file.each_line do |line|
    words = line.split

    # Separate each line into proper book attributes
    for i in 0..words.length-1 do
      words[i] = words[i].chomp(',')
    end

    # Add books to array
    book = Book.new(words[0], words[1], words[2], words[3], words[4])
    bookList.push(book)
  end
end

# Prompt user to pick sorting criteria
puts "Select what to sort the books by (isbn or title):"
sort = gets.chomp

# Check input validity
if sort != "isbn" && sort != "title"
  puts "Invalid sorting criteria, please try again"
  return
end

if sort == "isbn" then

  # Prompt user to select sorting direction
  puts "Select what direction to sort in (ascending or descending):"
  direction = gets.chomp

  if direction == "ascending" then
    sortBooks(bookList) {|a, b| a.isbn < b.isbn}
    
  elsif direction == "descending" then
    sortBooks(bookList) {|a, b| a.isbn > b.isbn}
    
  else
    puts "Invalid direction, please try again"
  end

elsif sort == "title" then
  sortBooks(bookList) {|a, b| a.title < b.title}

else
  puts "Invalid sorting criteria, please try again"
end

File: Project Part 2/TaskB.scala
import scala.io.Source
import java.io.File
import scala.io.StdIn.readLine
import java.io.PrintWriter

// Defintion of Book class with necessary attributes
class Book (val isbn: Long, val author: String, val title: String, val publisher: String)

// Definiton of sorting function that takes an array of Books and a function as parameters
def sortBooks(books: Array[Book], block: (Book, Book) => Boolean): Unit = {

  // Nested for loop for bubble sort using the provided function
  for (i <- 0 to books.size - 1) {
    for (j <- i + 1 to books.size - 1) {
      if (block(books(i), books(j)) == false) {
        val temp = books(i)
        books(i) = books(j)
        books(j) = temp
      }
    }
  }

  // Write sorted list of books to the file
  val writer = new PrintWriter("textfile.txt")
  for (i <- 0 to books.size - 1) {
    writer.write(books(i).isbn + ", " + books(i).author + ", " + books(i).title + ", " + books(i).publisher + "\n")
  }
  writer.close()
}

// Declare arrays and file reader necessary to get the list of books from the file
var bookList = Array[Book]()
var words = Array[String]()
val file = Source.fromFile("textfile.txt")
val lines = file.getLines()

// For loop to read all the lines in the file
for (line <- lines) {
  words = line.split(" ")

  // Format words correctly before adding to the book list
  for (i <- 0 to words.size - 1) {
    if (words(i).contains(",")) {
      words(i) = words(i).dropRight(1)
    }
  }
  var author = words(1) + " " + words(2)

  // Add books to the list
  var book = new Book(words(0).toLong, author, words(3), words(4))
  bookList = bookList :+ book
}

// Prompt user to select sorting criteria
println("Select what to sort the books by (isbn or title):")
val sort = readLine()

// Check input given by user
if (sort == "isbn") {

  // Prompt user to select sorting direction
  println("Select what direction to sort in (ascending or descending):")
  val direction = readLine()

  // Check input and call sorting function accordingly
  if (direction == "ascending") {
    sortBooks(bookList, {(a: Book, b: Book) => a.isbn < b.isbn})
  }
  else if (direction == "descending") {
    sortBooks(bookList, {(a: Book, b: Book) => a.isbn > b.isbn})
  }
  else {
    println("Invalid direction, please try again")
  }
}
else if (sort == "title") {
  sortBooks(bookList, {(a: Book, b: Book) => a.title < b.title})
}
else {
  println("Invalid sorting criteria, please try again")
}